File: frontend/packages/ui/src/stores/editMessageStore.ts
/**
 * editMessageStore.ts — Global state for the "Edit message" feature.
 *
 * When a user chooses "Edit" from the message context menu, this store
 * holds the editing context so that ChatHistory can dim messages,
 * MessageInput can populate the editor and show a banner, and
 * sendHandlers can delete messages from the edit point before re-sending.
 *
 * Architecture: docs/architecture/chat-edit-reprocess.md (pending)
 */

import { writable, get } from 'svelte/store';

export interface EditMessageState {
  /** Chat ID the edited message belongs to */
  chatId: string;
  /** The user message being edited */
  messageId: string;
  /** Original markdown content of the message */
  messageContent: string;
  /** created_at timestamp of the edited message — messages at/after this are dimmed */
  createdAt: number;
}

export const editMessageStore = writable<EditMessageState | null>(null);

/**
 * Enter edit mode for a specific user message.
 * Populates the store so ChatHistory dims messages and MessageInput loads content.
 */
export function startEdit(chatId: string, messageId: string, messageContent: string, createdAt: number): void {
  editMessageStore.set({ chatId, messageId, messageContent, createdAt });
  console.debug('[editMessageStore] Edit started:', { chatId, messageId, createdAt });
}

/**
 * Exit edit mode without sending. Restores all messages to normal.
 */
export function cancelEdit(): void {
  const current = get(editMessageStore);
  if (current) {
    console.debug('[editMessageStore] Edit cancelled for message:', current.messageId);
  }
  editMessageStore.set(null);
}

File: apple/OpenMates/Sources/Core/Services/EditMessageStore.swift
// Edit message store — tracks the message currently being edited in a chat.
// Mirrors the web app's editMessageStore.ts: stores the editing context so
// the input field can load the original content, and the message list can
// dim other messages during editing.

import Foundation
import SwiftUI

@MainActor
final class EditMessageStore: ObservableObject {
    static let shared = EditMessageStore()

    @Published private(set) var editingContext: EditContext?

    struct EditContext: Equatable {
        let chatId: String
        let messageId: String
        let originalContent: String
        let createdAt: String
    }

    var isEditing: Bool { editingContext != nil }
    var editingMessageId: String? { editingContext?.messageId }
    var editingChatId: String? { editingContext?.chatId }

    private init() {}

    func startEdit(chatId: String, messageId: String, content: String, createdAt: String) {
        editingContext = EditContext(
            chatId: chatId,
            messageId: messageId,
            originalContent: content,
            createdAt: createdAt
        )
    }

    func cancelEdit() {
        editingContext = nil
    }

    func completeEdit() {
        editingContext = nil
    }

    func isMessageBeingEdited(_ messageId: String) -> Bool {
        editingContext?.messageId == messageId
    }

    func isInEditMode(chatId: String) -> Bool {
        editingContext?.chatId == chatId
    }
}
